harness/asyncHelpers.js

```diff
@@ -0,0 +1,122 @@
+// Copyright (C) 2022 Igalia, S.L. All rights reserved.
+// This code is governed by the BSD license found in the LICENSE file.
+/*---
+description: |
+    A collection of assertion and wrapper functions for testing asynchronous built-ins.
+defines: [asyncTest]
+---*/
+
+function asyncTest(testFunc) {
+  if (!Object.hasOwn(globalThis, "$DONE")) {
+    throw new Test262Error("asyncTest called without async flag");
+  }
+  if (typeof testFunc !== "function") {
+    $DONE(new Test262Error("asyncTest called with non-function argument"));
+    return Promise.resolve();
+  }
+  try {
+    return testFunc().then(
+      function () {
+        $DONE();
+      },
+      function (error) {
+        $DONE(error);
+      }
+    );
```

Comment:
what happens if this is trying to test Promise.prototype.then itself?

Reply:
Good question. Looking at those tests, many of them wouldn't benefit from this function. But also, Promise.prototype.then seems core enough to async functionality that it seems there's a bit of a circular dependency to using both; perhaps that's one of the places it makes sense to continue to use custom logic around manually invoking $DONE? Very open to other ideas, if you have suggestions for how to improve this case.

Either way, I think it remains useful for simplifying tests of other async features not implementing core async interfaces.

Reply:
Thinking further about it: This only depends on the test function returning a thenable. There are a few options:
1. Do not use asyncTest when testing Promise.prototype.then. Instead, write your own logic for calling $DONE.
2. In your test for Promise.prototype.then, return a custom thenable rather than a Promise. This seems reasonable for a test of Promises anyway.
3. If testing a polyfilled Promise.prototype.then, one would expect the regular tests for Promise.prototype.then to uncover bugs, or the consistency tests for this harnenss to catch them. This itself is not a test for Promise.prototype.then, and doesn't reference or depend on it in any way, as it only references a thenable.

I'll push a quick update removing the references to Promise.resolve and instead returning a simple thenable, so that now the implementation doesn't reference Promise at all.

Does that cover all the cases you were thinking of, @ljharb ?